copier/vcs.py

```diff
@@ -108,23 +112,73 @@ def checkout_latest_tag(local_repo: StrOrPath, use_prereleases: OptBool = False)
         return latest_tag
 
 
-def clone(url: str, ref: OptStr = None) -> str:
-    """Clone repo into some temporary destination.
+def _perform_clone(src: str, dst: str, ref: OptStr = None) -> str:
+    """Clone from src into dst.
 
     Args:
-        url:
-            Git-parseable URL of the repo. As returned by
-            [get_repo][copier.vcs.get_repo].
+        src:
+            Git-parseable repo clone source.
+        dst:
+            Git-parseable repo clone destination.
         ref:
             Reference to checkout. For Git repos, defaults to `HEAD`.
     """
-    location = tempfile.mkdtemp(prefix=f"{__name__}.clone.")
-    _clone = git["clone", "--no-checkout", url, location]
+    _clone = git["clone", "--no-checkout", src, dst]
     # Faster clones if possible
     if GIT_VERSION >= Version("2.27"):
         _clone = _clone["--filter=blob:none"]
     _clone()
-    with local.cwd(location):
+
+    with local.cwd(dst):
         git("checkout", ref or "HEAD")
         git("submodule", "update", "--checkout", "--init", "--recursive", "--force")
-    return location
+
+    return dst
+
+
+def clone(url: str, ref: OptStr = None) -> str:
+    """Clone repo into some temporary destination.
+
+    Includes dirty changes for local templates by copying into a temp
+    directory and applying a wip commit there.
+
+    Args:
+        url:
+            Git-parseable URL of the repo. As returned by
+            [get_repo][copier.vcs.get_repo].
+        ref:
+            Reference to checkout. For Git repos, defaults to `HEAD`.
+    """
+
+    location = tempfile.mkdtemp(prefix=f"{__name__}.clone.")
+
+    if Path(url).is_dir():
+        is_dirty = False
+        with local.cwd(url):
+            is_dirty = bool(git("status", "--porcelain").strip())
+        if is_dirty:
+            url_abspath = Path(url).absolute()
+            with TemporaryDirectory(prefix=f"{__name__}.dirty.") as src:
+                with local.cwd(src):
+                    for item in os.listdir(url_abspath):
+                        s = os.path.join(url_abspath, item)
+                        d = os.path.join(src, item)
+                        if os.path.isdir(s):
+                            copytree(s, d)
+                        else:
+                            copy2(s, d)
```

Comment:
This turned out to be not so simple for a few reasons:
- `git diff` does not, by default, give a diff of the untracked files (of interested here doing a dirty copy)
- `HEAD` might not exist (e.g., `git init` was just run)

That said, I was able to get something closer to what you are asking for working by using `git diff --no-index` and then excluding `*/.git/*` on the apply. This will of course copy over other unwanted files to the temporary clone directory, but they shouldn't make it into the subdirectory since they will be in the clone's .gitignore. I think this is arguably just as inefficient in the worst-case (large .gitignored file getting needlessly copied).

I tried getting something like ` git --git-dir=<clone_dst>/.git --work-tree=<template_src> diff HEAD` [SO source](https://stackoverflow.com/a/32840462/916969) to work to no avail. Seems that anytime `git-diff` is used without `--no-index`, it ignores untracked files.

Need to make a couple fixes for CI still, but pushed latest changes to avoid losing progress.

Reply:
AFAIK `git stash` works both with staged and untracked files. Have you tried with it?